### ISSUE TITLE
cannot expect empty exception message

### ISSUE THREAD

**Issue by keradus** [sebastianbergmann/phpunit#2731]
| Q                   | A
| --------------------| ---------------
| PHPUnit version     | 5.7.21, but probably lowers as well
| PHP version         | 7.1.1-1, but not related
| Installation Method | Composer

I do know it's an edge case, but today I wanted to ensure that exception was raised with **empty** message:
```
$tihs->expectExceptionMessage('');
```

sadly, it's not working - it behaves like method is not called at all.

reason is here:
https://github.com/sebastianbergmann/phpunit/blob/5.7/src/Framework/TestCase.php#L1089

setting up the value with `null` and with empty string are treated the same way.
IMO, only null shall be checked there

**Comment by wesolowski** [comment]
i can't reproduce them. This code work fine:
```php
<?php

class Example
{
    public function foo()
    {
        throw new \Exception('');
    }
}

class ExampleTest extends PHPUnit_Framework_TestCase
{
    public function testFoo()
    {
        $this->expectExceptionMessage('');
        (new \Example())->foo();
    }
}
```

Console Output:
![screen](https://user-images.githubusercontent.com/1388643/31544517-75dc9bd0-b01b-11e7-828a-7264d7b527f0.png)

---

But i find another error.
This Code don't throws any error.

```php
<?php

class Example
{
    public function foo()
    {
        throw new \Exception('One');
    }
}

class ExampleTest extends PHPUnit_Framework_TestCase
{
    public function testFoo()
    {
        $this->expectExceptionMessage('');
        (new \Example())->foo();
    }
}
```

**Comment by keradus** [comment]
it's not throwing error because `expectExceptionMessage` expect exception to be raised (configured previously via `expectException` or any `\Exception` if not), and then msg is not checked as I described

**Comment by sebastianbergmann** [comment]
`expectExceptionMessage()` etc. can be used without using `expectException()`:

```php
<?php
class Test extends PHPUnit\Framework\TestCase
{
    public function testOne()
    {
        $this->expectExceptionMessage('message');
    }
}
```
Before 6727744434bf19016364aef1a49f80b32d045b8b the above would result in
```
PHPUnit 6.4.1-5-ge0bc901f2 by Sebastian Bergmann and contributors.

F                                                                   1 / 1 (100%)

Time: 33 ms, Memory: 4.00MB

There was 1 failure:

1) Test::testOne
Failed asserting that exception of type "Exception" is thrown.

FAILURES!
Tests: 1, Assertions: 1, Failures: 1.
```
I found that confusing and have changed it to
```
PHPUnit 6.4.1-6-g672774443 by Sebastian Bergmann and contributors.

F                                                                   1 / 1 (100%)

Time: 46 ms, Memory: 4.00MB

There was 1 failure:

1) Test::testOne
Failed asserting that exception with message "message" is thrown

FAILURES!
Tests: 1, Assertions: 1, Failures: 1.
```
now.

**Comment by sebastianbergmann** [comment]
I probably/subconsciously fixed this issue as part of 6727744434bf19016364aef1a49f80b32d045b8b.

**Comment by keradus** [comment]
@sebastianbergmann is 5.x line closed? bug is there since 5.x

**Comment by sebastianbergmann** [comment]
I'll look into backporting this to PHPUnit 5.7 (which still gets bug fixes until February 2018).

**Comment by keradus** [comment]
damn, it's hard to track changes when they are commited directly to branch... :(

https://github.com/sebastianbergmann/phpunit/commit/567ca23a4c62e8b9bcefae0b39b5f33e3853c2c2#commitcomment-24975103

**Comment by remicollet** [comment]
Notice that empty exception can be test without any change

`$this->expectExceptionMessageRegExp('/^$/')`

This fix in 5.7 introduce a **BC break** in how `$this->expectExceptionMessage('')` is handled.

See
* https://github.com/rmccue/Requests/issues/295
* https://github.com/zendframework/zend-mail/issues/173
